smurfFinalProject/smurfFinalProject/main.cpp

```diff
@@ -0,0 +1,198 @@
+node *assign(const SemanticValues &sv){                             //the same structure, but instead of an operation node as a parent,
+    node *left = sv[0].get<ParseTreeNode>().get();                  //the parent node is an assignment node and the left child is a
+    node *right = sv[1].get<ParseTreeNode>().get();                 //variable or identifier:
+    left = new assignmentNode(left, right);                         //              assignment
+    return left;                                                    //             |          |
+};                                                                  //        variable        value
+    
+node *code(const SemanticValues &sv){                               //The code function allows for multiple statements to be read in from
+    codeNode *code = new codeNode();                                //the command line. It takes in the strings passed to the parser and
+    for(int i=0; i<sv.size(); i++){                                 //adds each statement to a child node under the main codeNode in a vector.
+        node *x = sv[i].get<ParseTreeNode>().get();                 //It can take in as many children statement nodes as it needs to.
+        code->addToVect(x);                                         //              c  o  d  ... e
+    }                                                               //              |    |   ... |
+    return code;                                                    //      statement statement  statement
+};
+
+node *statement(const SemanticValues &sv){                          //The statement function allows for entire statements that may consist of
+    node *x = sv[0].get<ParseTreeNode>().get();                     //multiple nodes to be compressed into one so that it can be passed to the
+    return x;                                                       //codeNode's vector for storage before interpretation.
+}
+
+void setup_ast_generation(parser &parser)
+{
+    parser["code"] = [](const SemanticValues &sv){
+        return ParseTreeNode(code(sv));
+    };
+    
+    parser["statement"] = [](const SemanticValues &sv) {
+        node *n = statement(sv);
+        return ParseTreeNode(n);
+    };
+    
+    parser["let_stmt"] = [](const SemanticValues &sv) {
+        return sv[0];
+    };
+    
+    parser["decl"] = [](const SemanticValues &sv) {
+        node *n = assign(sv);
+        return ParseTreeNode(n);
+    };
+    
+    parser["assignment"] = [](const SemanticValues &sv) {
+        node *n = assign(sv);
+        return ParseTreeNode(n);
+    };
+    
+    parser["identifier"] = [](const SemanticValues &sv) {
+        return ParseTreeNode(new variableNode(sv.token()));
+    };
+    
+    parser["expr"] = [](const SemanticValues &sv) {
+        node *n = bin_op(sv);
+        return ParseTreeNode(n);
+    };
```

Comment:
An expr isn't a binop. This still works: can you work out why?

But this clause, and the let_stmt clause above, actually do nothing; your parser works fine if I remove them both.

# -2